packages/context/src/provider.ts

```diff
@@ -0,0 +1,89 @@
+// Copyright IBM Corp. 2013,2017. All Rights Reserved.
+// Node module: loopback
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {Context} from './context';
+import {BoundValue} from './binding';
+import {Constructor, instantiateClass} from './resolver';
+
+/**
+ * Provider of a value
+ */
+export interface Provider {
+  value<T>(): T;
+}
+
+/**
+ * BindingProvider provisions injecting a value into the binding asynchronously
+ * @example:
+ * ```ts
+ * ctx.bind('provider_key').toProvider(MyProvider);
+ * ```
+ */
+export abstract class BindingProvider<T> implements Provider {
+  abstract value(): T;
+}
+
+/**
+ * ValueProvider injects a value into the binding asynchronously
+ * @example:
+ * ```ts
+ * export class MyProvider extends ValueProvider {
+ *   constructor(@inject('value') param){}
+ *   value<string>(): string {
+ *     return computed_value;
+ *   }
+ * }
+ * ```
+ */
+export abstract class ValueProvider implements BindingProvider<BoundValue | Promise<BoundValue>> {
+  abstract value(): BoundValue | Promise<BoundValue>;
+}
+
+/**
+ * FunctionProvider injects a function into the binding asynchronously
+ * @example:
+ * ```ts
+ * export class MyProvider extends FunctionProvider {
+ *   constructor(@inject('value') param){}
+ *   value<() => string>(): () => string {
+ *     if (param)
+ *        return function1
+ *     else
+ *        return function2;
+ *   }
+ * }
+ * ```
+ */
+export abstract class FunctionProvider implements BindingProvider<() => BoundValue | Promise<BoundValue>> {
+  abstract value(): () => BoundValue | Promise<BoundValue>;
```

Comment:
() => BoundValue | Promise<BoundValue> is the factory function, which concurs with the definition in binding.ts